### PR TITLE
[Hail] Fix several CSE bugs

### LINKED ISSUE #7418: Python CSE bug
I'm not sure what the right thing is here but this seems wrong:

```python
In [4]: ht = hl.utils.range_table(10)

In [5]: x = hl.agg.count()

In [6]: ht.aggregate((x, hl.agg.filter(ht.idx % 2 == 0, x)))
Out[6]: (10, 10)
```

### PR DESCRIPTION
**Pull request by patrick-schultz**
fixes #7418

The main change here is fixing #7418 by making nodes which perform aggregation aware of the nearest containing node which defines what is being aggregated over. This is done by making a new variable name---here called "agg_capability"---which is added to the child context by any node which (re)defines the meaning of aggregation (`TableMapRows`, `AggFilter`, etc.), and which is implicitly referenced by any node which performs an aggregation (`ApplyAggOp`, `AggFilter`, etc.).

This requires nodes to be able to bind variables which shadow variables already bound by parents, which it turns out wasn't handled correctly by the CSE algorithm. Fixing this required several changes:
* I moved free variable computation to a lazy value on `BaseIR`. This way, each time we see a subtree `x`, we can recompute the max depth of the binding sites of all of `x`'s free variables, since those binding sites may be different than last time we saw `x`. This also required splitting the free variables into value, agg, and scan sets, so they can be looked up in the correct context (previously context lookup was always done at the `Ref` node, at which point the variable was in the value context).
* Now, in the `CSEPrintPass`, we have to recompute the same binding depth calculation that was done in the analysis pass, so we know which binding site to look at (previously I just searched all binding sites in scope, but with shadowing handled correctly we don't have sufficient information to decide which binding site is valid). This requires maintaining contexts in the print pass, which is annoying because we are now traversing the tree of `Renderable` children, which is not exactly the same as the IR tree.
* To fix this, I duplicated all methods involving binding structure on `BaseIR`. To avoid having to write twice as many methods on concrete IR classes, I made the methods taking the index of the `Renderable` child (e.g. `renderable_bindings`) be the primary methods which are overridden. For all IR nodes, there is a map from IR child index to Renderable child index, defined in `renderable_idx_of_child`, which is used to define the `bindings` and similar methods in terms of the `renderable_bindings` and similar. This is messier than I would have liked, but I couldn't think of a better way.

### CODE REVIEW

File: hail/python/hail/ir/base_ir.py [hunk 1]
```diff
@@ -204,6 +243,41 @@ def parse(self, code, ref_map={}, ir_map={}):
             {k: t._parsable_string() for k, t in ref_map.items()},
             ir_map)
 
+    @property
+    def free_vars(self):
+        def vars_from_child(i):
+            if self.uses_agg_context(i):
+                return self.children[i].free_agg_vars
```
Comment (tpoterba) [inline hail/python/hail/ir/base_ir.py]: (let's dig into what's going on here, as we talked about)

**Comment by tpoterba** [comment]
just a note that we're waiting for this PR to go in for the 0.2.27 release.

File: hail/python/hail/ir/renderer.py [hunk 2]
```diff
@@ -654,47 +591,70 @@ def add_lets(self, let_bodies: Sequence[str], out_builder: MutableSequence[str])
...
-                             min_binding_depth: int,
-                             min_value_binding_depth: int,
-                             scan_scope: bool,
-                             depth: int):
+                             bindings_stack: 'Dict[int, CSEPrintPass.BindingsStackFrame]'):
+            child_min_binding_depth = self.min_binding_depth
+            child_min_value_binding_depth = self.min_value_binding_depth
+            child_scan_scope = self.scan_scope
+
+            # TODO: can this be moved into make_child_frame?
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/ir/renderer.py]: 🤦‍♂

**Comment by tpoterba** [comment]
lots of test failues

**Pull request merged**